### PR TITLE
Swat/feature/17314 enhancing error logging

### PR DESCRIPTION
**Pull request by sshankar19**
Request ID generation + logging
PostServerErrorCallback
Propogation of Client specific logs to the backend
Logging amorphic version at startup

Replaces: https://github.com/haven-life/amorphic/pull/49

### CODE REVIEW

File: lib/listen.js [hunk 1]
```diff
@@ -12,6 +12,31 @@ let SupertypeSession = require('@havenlife/supertype').SupertypeSession;
...
+const packageVersions = resolveVersions([
+	'@havenlife/semotus',
+	'@havenlife/supertype',
+	'@havenlife/persistor',
+	'@havenlife/bindster'
+]);
+
+packageVersions['amorphic'] = require('../../package.json').version;
+
+function resolveVersions(packages) {
```
Comment (nsaenz7) [inline lib/listen.js:24]: what's the need to add custom module resolution logic? this seems like something if we can avoid doing ourselves we should

Comment (sshankar19) [reply lib/listen.js:24]: This doesn't exist, currently, and 
'require.resolve(dependency)', doesn't actually `require` or import the package. It simply returns the filepath of the package, and if there's an issue (like it's not there), it throws an error. 

We're just checking for the existence of the package, and then if it exists, pulling out the version. Believe it or not, this doesn't exist in node!

**Review by srksag-zz — DISMISSED** [review]

File: client.js [hunk 2]
```diff
@@ -432,8 +432,8 @@ amorphic = // Needs to be global to make mocha tests work
 
     generateUID: function () {
         // shamelessly taken from https://stackoverflow.com/questions/8012002/create-a-unique-number-with-javascript-time/28918947
-        this.requestUID = new Date().valueOf().toString(36) + Math.random().toString(36).substr(2);
-        return this.requestUID;
+        this.storedUID = new Date().valueOf().toString(36) + Math.random().toString(36).substr(2);
+        return this.storedUID;
```
Comment (srksag-zz) [inline client.js:436]: I know this should not cause any issues, but for the apps running on multiple containers, we can still get the same id, right?

Comment (nsaenz7) [reply client.js:436]: that's a good point - assuming any of the apps can produce the same date string, this is only as unique as the math.random is...this isn't for any data management though this is just tracing logs through, right? so conflicts won't necessarily kill us here, assuming they're going to be pretty rare to begin with

Comment (sshankar19) [reply client.js:436]: Yes, but bear in mind, that once within an app, we're guaranteed virtual uniqueness if we pair this requestId with the already existing sessionId that we have access to in the logger's context

**Review by nsaenz7 — APPROVED** [review]

**Review by srksag-zz — APPROVED** [review]

**Pull request merged**